File: core/datacap-spi/src/main/java/io/edurt/datacap/spi/executor/ExecutorManager.java
package io.edurt.datacap.spi.executor;

import com.google.inject.AbstractModule;
import lombok.extern.slf4j.Slf4j;

import java.util.ServiceLoader;

@Slf4j
public class ExecutorManager
        extends AbstractModule
{
    private final Iterable<ExecutorModule> externalModules;

    public ExecutorManager()
    {
        this.externalModules = ServiceLoader.load(ExecutorModule.class);
    }

    public ExecutorManager(Iterable<ExecutorModule> externalModules)
    {
        this.externalModules = externalModules;
    }

    @Override
    protected void configure()
    {
        log.info("Loading executor start ...");
        for (ExecutorModule plugin : this.externalModules) {
            log.info("Install executor <{}>", plugin.getClass().getSimpleName());
            this.install(plugin);
        }
        log.info("Loading executor end ...");
    }
}

File: executor/datacap-executor-spi/src/main/kotlin/io/edurt/datacap/executor/ExecutorManager.kt
package io.edurt.datacap.executor

import com.google.inject.AbstractModule
import org.slf4j.LoggerFactory.getLogger
import java.time.LocalDateTime
import java.util.*

class ExecutorManager : AbstractModule {
    private val log = getLogger(this.javaClass)
    private var externalModules: Iterable<ExecutorModule>? = null

    constructor() {
        this.externalModules = ServiceLoader.load(ExecutorModule::class.java)
    }

    override fun configure() {
        log.info("================ Executor started ================")
        externalModules !!.forEach { module ->
            log.info("Install Executor [ {} ] Join time [ {} ]", module.name(), LocalDateTime.now())
            this.install(module)
        }
        log.info("================ Executor end ================")
    }
}
